examples/diviner/train.py

```diff
@@ -0,0 +1,120 @@
+    :return: A Pandas DataFrame that is a concatenation of each entry of the location_data tuple
+             values with a datetime column added via hourly intervals for the duration of each
+             series. The original data is in hourly electrical consumption values.
+    """
+    raw_data = datasets.load_taylor(as_series=True)
+    start = datetime.strptime(start_dt, "%Y-%m-%d %H:%M:%S")
+    dates = pd.date_range(start=start, periods=len(raw_data), freq="H").values
+
+    generated_listing = []
+
+    for (country, city) in location_data:
+        generated_listing.append(
+            pd.DataFrame(
+                {
+                    "watts": np.random.uniform(0.3, 0.8) * raw_data,
+                    "datetime": dates,
+                    "country": country,
+                    "city": city,
+                }
+            )
+        )
+
+    output = pd.concat(generated_listing).reset_index().drop("index", axis=1)
+    return output
+
+
+def grouped_prophet_example(locations, start_dt, artifact_path):
+
+    print("Generating data...\n")
+    data = generate_data(location_data=locations, start_dt=start_dt)
+    grouping_keys = ["country", "city"]
+    print("Data Generated.\nBuilding GroupedProphet Model...")
+
+    model = GroupedProphet(n_changepoints=96, uncertainty_samples=0).fit(
+        df=data, group_key_columns=grouping_keys, y_col="watts", datetime_col="datetime"
+    )
+    print("GroupedProphet model built.\n")
+
+    params = model.extract_model_params()
+
+    print(f"Params: \n{params.to_string()}")
+
+    print("Running Cross Validation on all groups...\n")
+    metrics = model.cross_validate_and_score(
+        horizon="120 hours",
+        period="480 hours",
+        initial="960 hours",
+        parallel="threads",
+        rolling_window=0.05,
+        monthly=False,
+    )
+    print(f"Cross Validation Metrics: \n{metrics.to_string()}")
+
+    mlflow.diviner.log_model(diviner_model=model, artifact_path=artifact_path)
+
+    # As an Alternative to saving metrics and params directly with a `log_dict()` function call,
+    # Serializing the DataFrames to local as a .csv can be done as well, without requiring
+    # column or object manipulation as shown below this block:
+
+    local = "/tmp/diviner/data"
+    os.makedirs(local, exist_ok=True)
+    params.to_csv(f"{local}/params.csv", index=False, header=True)
+    metrics.to_csv(f"{local}/metrics.csv", index=False, header=True)
+    mlflow.log_artifacts(local, artifact_path="run_data")
+
+    # Saving the parameters and metrics as json without having to serialize to local
+    # NOTE: this requires casting of fields that cannot be serialized to JSON
+    # NOTE: Do not use both of these methods. These are shown as an either/or alternative based
+    # on how you would choose to consume, view, or analyze the per-group metrics and parameters.
+
+    params["t_scale"] = params["t_scale"].astype(str)  # NB: Cannot serialize a Timestamp as JSON
```

Comment:
This is only a problem with Prophet, which is why I included this 'painful' example here. pmdarima doesn't have objects stored in the param extraction.

Reply:
a most excellent idea! 
converted in the simplest way possible (similar to what Pandas does with `.to_string()` and `.to_csv()`):

``` python
params = params.astype(dtype=str, errors="ignore")
```
to cast the entire df to str type and to ignore errors so that examples don't blow up.